C51-53: Implement Bulk actions for case list

swastikpareek opened the pull request:
# Overview
This PR contains Bulk actions feature for case list page with new designs 

## Before 
<img width="1680" alt="c51-53 - before" src="https://user-images.githubusercontent.com/3340537/46871546-6199ae80-ce4f-11e8-8a42-927c79036f50.png">

## After
![c51-53- bulk actions after](https://user-images.githubusercontent.com/3340537/46872085-d6211d00-ce50-11e8-838b-f6478b9b8960.gif)


## Technical Details
* Created a reusable component `civicaseBulkActionsChecboxes` and `civicaseBulkActionsMessages` which uses the same controller.
* Every thing is handled by event subscribers and emitters. The following events are newly created
  * `civicase::bulk-actions::check-box-toggled` - triggers when bulk actions checkbox is clicked
  * `civicase::bulk-actions::bulk-selections ` - Fires when any bulk selection is clicked (Everything, visible, None)
  * `civicase::case-list::header-position-changed` - should be fired whenever the sticky case list headers position needs to be readjusted
  * `civicase::case-details::header-position-changed ` - should be fired whenever the sticky case summary tab header needs to readjusted.
  * `civicase::case-details::filter-position-changed` - should be fired whenever the sticky case detail filter header is to be readjusted.
 * Event subscribers are added in `caseListTable.js` for deselect/ select all/ select visible . Same can be referred for implementing bulk actions at other places.
* BEM classes used 
  * .civicase__bulkactions-checkbox
  * .civicase__bulkactions-checkbox-toggle
  * .civicase__bulkactions-select-mode-dropdown
  * .civicase__bulkactions-actions-dropdown 
  * .civicase__bulkactions-message 
  * .civicase__link-disabled - for disabling link with pointer-events set to none.
* Written tests for each part.

## Updates in d866d4ea
### Problem
When selecting a case and deselecting it again. the message box appears with 0 selections.

### Solution

 This was coming because the `bulkActionMessages` directive has `ng-hide` applied to itself in `caseListTable.html ` and had ng-show in its template. Since in the directive we were using `replace: true` both the conditions to the same div which sometimes incorrectly adds `ng-hide` class to the div.
`caseListTable.html`
```
<civicase-bulk-actions-message selected-items="selectedCases.length" total-count="totalCount" is-select-all-available="isSelectAllAvailable" show-checkboxes="showCheckboxes" ng-hide="viewingCase"></civicase-bulk-actions-message>
```
and `BulkActionsMessage.html`
```
<div class="civicase__bulkactions-message" ng-show="selectedItems > 0 && showCheckboxes">
```
in `BulkActionsMessage.js`
```
(function (angular, $, _) {
  var module = angular.module('civicase');

  module.directive('civicaseBulkActionsMessage', function () {
    return {
      restrict: 'EA',
      replace: true,
      controller: 'civicaseBulkActionsController',
      templateUrl: '~/civicase/BulkActionsMessage.html',
      scope: {
        selectedItems: '=',
        isSelectAllAvailable: '=',
        totalCount: '=',
        showCheckboxes: '='
      }
    };
  });
})(angular, CRM.$, CRM._);
```


So  removing `replace: true` from the directive fix this as now the ng-show from inside and ng-hide from outside will be applied on different div eliminating the chance of ambiguity. Updated the tests as well.

deb1990 commented:
![2018-10-15 at 9 11 am](https://user-images.githubusercontent.com/5058867/46928871-4cdd3680-d05a-11e8-871d-40530123f169.png)

1. Actions text should not be ALL CAPS, also font color does not match design.
2. In Case Viewing mode, the Bulk Actions should be visible according to the design.

![2018-10-15 at 9 13 am](https://user-images.githubusercontent.com/5058867/46928940-9cbbfd80-d05a-11e8-8ed5-f8a8378d5a06.png)
As per discussion with Shitij, count of `Everything`, `Displayed` should be shown in brackets.

deb1990 commented:
![2018-10-15 at 9 22 am](https://user-images.githubusercontent.com/5058867/46929170-d3dede80-d05b-11e8-8b8e-affe74eeaa13.png)
@swastikpareek how about having this row below the header of the table? could you check with Jon please.

deb1990 commented on ang/civicase/BulkActionsMessage.html:
```diff
@@ -0,0 +1,5 @@
+<div class="civicase__bulkactions-message" ng-show="selectedItems > 0 && showCheckboxes">
+  <div class="alert">
+    You have selected <strong>{{selectedItems}}</strong> <ng-pluralize count="selectedItems" when="{'one': 'item', 'other': 'items'}"></ng-pluralize>. <a ng-show="totalCount !== selectedItems" href="#" ng-class="{'civicase__link-disabled': !isSelectAllAvailable}" ng-click="select('all'); $event.stopPropagation(); $event.preventDefault();">Select all {{totalCount}} that match the search</a> <a ng- ...
```
Dont use `strong`, use `span` with css instead.

deb1990 replied:
Sorry its fine to use `strong`.

deb1990 commented on ang/civicase/BulkActionsMessage.html:
```diff
@@ -0,0 +1,5 @@
+<div class="civicase__bulkactions-message" ng-show="selectedItems > 0 && showCheckboxes">
+  <div class="alert">
+    You have selected <strong>{{selectedItems}}</strong> <ng-pluralize count="selectedItems" when="{'one': 'item', 'other': 'items'}"></ng-pluralize>. <a ng-show="totalCount !== selectedItems" href="#" ng-class="{'civicase__link-disabled': !isSelectAllAvailable}" ng-click="select('all'); $event.stopPropagation(); $event.preventDefault();">Select all {{totalCount}} that match the search</a> <a ng- ...
```
Please dont have such long lines, its hard to read. break them down.
```
<div class="civicase__bulkactions-message" ng-show="selectedItems > 0 && showCheckboxes">
  <div class="alert">
    You have selected <strong>{{selectedItems}}</strong>
    <ng-pluralize count="selectedItems" when="{'one': 'item', 'other': 'items'}"></ng-pluralize>.
    <a
      ng-show="totalCount !== selectedItems" href="#"
      ng-class="{'civicase__link-disabled': !isSelectAllAvailable}"
      ng-click="select('all'); $event.stopPropagation(); $event.preventDefault();">
      Select all {{totalCount}} that match the search
    </a>
    <a
      ng-show="totalCount === selectedItems" href="#"
      ng-class="{'civicase__link-disabled': !isSelectAllAvailable}"
      ng-click="select('none'); $event.stopPropagation(); $event.preventDefault();">
      Clear All
    </a>
  </div>
</div>
```

deb1990 replied:
`$event.stopPropagation(); $event.preventDefault();` can be moved inside `select`.

deb1990 commented on ang/civicase/BulkActionsMessage.js:
```diff
@@ -0,0 +1,18 @@
+    return {
+      restrict: 'EA',
+      replace: true,
+      controller: 'civicaseBulkActionsController',
```
Why dont you just use `civicaseBulkActionsController` as `ng-controller` in the HTML and remove the JS file. Would that work?

swastikpareek replied:
This can't be done for following reasons
* This is a directive and if we want to make the HTML reusable then its better to use it as a directive. if we use the above approach then we will have to re write the whole HTML.
* Also the attributes that are used in this directive are specific to itself and make it reusable. For e.g. `selectedItems`  for cases will have different variable name (selected-items="selectedCases.length") and for activities might be different (selected-items="selectedActivities.length"

deb1990 replied:
Lets discuss this.

deb1990 commented on ang/civicase/CaseCard--case-list.html:
```diff
@@ -4,15 +4,15 @@
   * Update links with activityFeedUrl function activityFeedUrl(caseid, activityType to link, activityStatus to Link)
   -->
-<div class="panel panel-default panel-strip-primary civicase__case-card civicase__animated-checkbox-card" ng-style="{'border-left-color': data.color}" ng-class="{'civicase__case-card--active': $parent.viewingCase === data.id, 'civicase__case-card--closed': data.status === 'Resolved'}" >
+<div class="panel panel-default panel-strip-primary civicase__case-card civicase__animated-checkbox-card" ng-style="{'border-left-color': data.color}" ng-class="{'civicase__case-card--active': $parent.viewingCase === data.id, 'civicase__case-card--closed': data.status === 'Resolved', 'civicase__animated-checkbox-card--expanded': $parent.showCheckboxes && !$parent.viewingCase && $parent.selectedCa ...
```
Break long lines please.

deb1990 commented on ang/civicase/CaseCard--case-list.html:
```diff
@@ -4,15 +4,15 @@
     </div> -->
-    <span class="civicase__checkbox" ng-if="!$parent.viewingCase && $parent.selectedCases && $parent.bulkAllowed" role="checkbox" aria-checked="{{activity.is_completed}}" tabindex="0" aria-labelledby="Mark complete/incomplete">
+    <span class="civicase__checkbox" ng-if="!$parent.viewingCase && $parent.selectedCases && $parent.bulkAllowed" role="checkbox" aria-checked="{{activity.is_completed}}" tabindex="0" aria-labelledby="Mark complete/incomplete" ng-click="toggleSelected(); $event.stopPropagation();">
       <i ng-if="data.selected" class="civicase__checkbox--checked material-icons">check_box</i>
```
Break long lines please.

deb1990 commented on ang/civicase/CaseCard--case-list.html:
```diff
@@ -4,15 +4,15 @@
-<div class="panel panel-default panel-strip-primary civicase__case-card civicase__animated-checkbox-card" ng-style="{'border-left-color': data.color}" ng-class="{'civicase__case-card--active': $parent.viewingCase === data.id, 'civicase__case-card--closed': data.status === 'Resolved'}" >
+<div class="panel panel-default panel-strip-primary civicase__case-card civicase__animated-checkbox-card" ng-style="{'border-left-color': data.color}" ng-class="{'civicase__case-card--active': $parent.viewingCase === data.id, 'civicase__case-card--closed': data.status === 'Resolved', 'civicase__animated-checkbox-card--expanded': $parent.showCheckboxes && !$parent.viewingCase && $parent.selectedCa ...
   <div class="panel-body">
     <!--
```
Do we still need this TODO?

deb1990 commented on ang/civicase/CaseCard--case-list.html:
```diff
@@ -4,15 +4,15 @@
   * use 'civicase__animated-checkbox-card--expanded' class : for enabled checkbox state
```
Do we still need this?

swastikpareek replied:
No we don't need this. Let me remove it.

deb1990 commented on ang/civicase/CaseCard--case-list.html:
```diff
@@ -4,15 +4,15 @@
   * Update links with activityFeedUrl function activityFeedUrl(caseid, activityType to link, activityStatus to Link)
   -->
-<div class="panel panel-default panel-strip-primary civicase__case-card civicase__animated-checkbox-card" ng-style="{'border-left-color': data.color}" ng-class="{'civicase__case-card--active': $parent.viewingCase === data.id, 'civicase__case-card--closed': data.status === 'Resolved'}" >
+<div class="panel panel-default panel-strip-primary civicase__case-card civicase__animated-checkbox-card" ng-style="{'border-left-color': data.color}" ng-class="{'civicase__case-card--active': $parent.viewingCase === data.id, 'civicase__case-card--closed': data.status === 'Resolved', 'civicase__animated-checkbox-card--expanded': $parent.showCheckboxes && !$parent.viewingCase && $parent.selectedCa ...
```
We should need to mention this many conditions.
`$parent.showCheckboxes && !$parent.viewingCase && $parent.selectedCases && $parent.bulkAllowed`

When `bulkAllowed` is false, `showCheckboxes` should be disabled by default. In fact the Bulk Allowed logic should be part of BulkActions directive. And unless user have permission, showCheckboxes should be disabled by default.

Also could you remind me, why `!$parent.viewingCase && $parent.selectedCases` is needed to be checked here?

swastikpareek replied:
> When bulkAllowed is false, showCheckboxes should be disabled by default. In fact the Bulk Allowed logic should be part of BulkActions directive. And unless user have permission, showCheckboxes should be disabled by default.

`bulkAllowed` is for permissions and show checkbox is for toggle view, both are different. However, this is a nice idea to make BulkActions to be a part of the directive. Let me see to this. and get back to you.

> Also could you remind me, why !$parent.viewingCase && $parent.selectedCases is needed to be checked here?

Sure. In the current design the checkbox doesn't show up when any case is in view state also if at least one case is selected, that means checkbox were enabled and if user navigates within pagers then the checkboxes should be visible. Making user know that he/she have some selection already in progress. Also if one of the cases is selected the actions dropdown will be available and if it is then the user should see the checkboxes for each cases throughout pagers.

deb1990 replied:
Lets discuss this.

deb1990 commented on ang/civicase/CaseCard.js:
```diff
@@ -15,12 +15,22 @@
     });
+
+    /**
+     * Update the checkbox the case card
```
There are two kings of checkbox in the case card, we need to be more specific.
`Update Bulk Action checkbox of the case card`

deb1990 commented on ang/civicase/CaseCard.js:
```diff
@@ -15,12 +15,22 @@
   });
 
   module.controller('CivicaseCaseCardController', function ($scope, getActivityFeedUrl, DateHelper) {
+    $scope.data.selected = $scope.data.selected || false;
```
If `$scope.data.selected` is not true, doe not that mean its already false? Why is this line required?

swastikpareek replied:
We needed this to initialise the variable as it was throwing some error for some condition. which I can't recall right now. However , removing this line don't show any impact on the functionality. So removing it.

deb1990 commented on ang/civicase/CaseCard.js:
```diff
@@ -15,12 +15,22 @@
+     */
+    $scope.toggleSelected = function () {
+      $scope.data.selected = !$scope.data.selected;
+      $scope.$emit('bulkSelectionCheckboxClicked', $scope.data);
```
Please follow this structure for emit messages.
`<appName>::<componentName/eventCategory>::<eventName>`
`'civicase::bulk-actions::check-box-toggled', $scope.activity);`

deb1990 commented on ang/civicase/CaseDetails.js:
```diff
@@ -311,12 +311,21 @@
+
+      $rootScope.$on('stickyCaseDetailsHeaderReinitiatePos', function () {
+        $timeout(function () {
+          // Resert right case view tab header
```
Typo in `Resert`

deb1990 commented on ang/civicase/Search.js:
```diff
@@ -83,22 +83,8 @@
-          $caseNavigation.data('bs.affix').options.offset.top = $casePanelBody.offset().top - bodyPadding;
-        }
-      });
+      $rootScope.$broadcast('stickyCaseListHeaderReinitiatePos');
```
Use `$emit`

deb1990 replied:
We dont need two events. Just fire a single event to denote, UI has changed. That should be listened by all.
For example, `civicase::case-list::header-position-changed`

swastikpareek replied:
> Use `$emit`

Limitation of using $emit is that it can be used in hierarchical way (i.e. any emitted event can only be caught by its parent scope or itself) but this event is being broadcasted from multiple controllers (Search.js) and (Bulkactionmessage.js) then we must need to use rootscope broadcast.

deb1990 replied:
Lets discuss this.

deb1990 commented on ang/civicase/CaseListDirectives.js:
```diff
@@ -31,9 +31,32 @@
+       */
+      function initSubscribers () {
+        $rootScope.$on('stickyCaseListHeaderReinitiatePos', function () {
+          $timeout(function () {
```
Why do we need `timeout` here?

swastikpareek replied:
We have to call the logic when DOM has updated and the best way to ensure that is to call the logic after the angular is done with its DOM updates. To make sure it is called after that we need to add a timeout. 

Please let me know if we need to discuss it further.

deb1990 replied:
Lets discuss this.

deb1990 commented on ang/civicase/CaseListTable.html:
```diff
@@ -7,16 +8,16 @@
-                <span class="caret"></span>
-              </button>
-              <ul class="dropdown-menu" civicase-actions="selectedCases" multiple="true" refresh-callback="refresh"></ul>
+            <!--  -->
```
missing doc

swastikpareek replied:
Its not missing doc but an empty inline comment missed . Let me remove it.

deb1990 commented on ang/civicase/CaseListTable.html:
```diff
@@ -7,16 +8,16 @@
-              </button>
-              <ul class="dropdown-menu" civicase-actions="selectedCases" multiple="true" refresh-callback="refresh"></ul>
+            <!--  -->
+            <civicase-bulk-actions-checkboxes ng-show="!viewingCase && selectedCases && bulkAllowed && !item.lock" class="pull-left" show-checkboxes="showCheckboxes" is-select-all-available="isSelectAllAvailable" selected-items="selectedCases.length"></civicase-bulk-actions-checkboxes>
```
Break long lines.

deb1990 commented on ang/civicase/CaseListTable.html:
```diff
@@ -7,16 +8,16 @@
-              <ul class="dropdown-menu" civicase-actions="selectedCases" multiple="true" refresh-callback="refresh"></ul>
+            <!--  -->
+            <civicase-bulk-actions-checkboxes ng-show="!viewingCase && selectedCases && bulkAllowed && !item.lock" class="pull-left" show-checkboxes="showCheckboxes" is-select-all-available="isSelectAllAvailable" selected-items="selectedCases.length"></civicase-bulk-actions-checkboxes>
+            <span class="civicase__bulkactions-actions btn-group btn-group-sm" ng-if="!viewingCase && selectedCases && bulkAllowed && !item.lock">
```
Same as the other comment
> We should need to mention this many conditions.
$parent.showCheckboxes && !$parent.viewingCase && $parent.selectedCases && $parent.bulkAllowed

> When bulkAllowed is false, showCheckboxes should be disabled by default. In fact the Bulk Allowed logic should be part of BulkActions directive. And unless user have permission, showCheckboxes should be disabled by default.

> Also could you remind me, why !$parent.viewingCase && $parent.selectedCases is needed to be checked here?

swastikpareek replied:
Please see my comments above

deb1990 commented on ang/civicase/CaseListTable.html:
```diff
@@ -7,16 +8,16 @@
+                  <span class="caret"></span>
+                </button>
+                <ul class="dropdown-menu" civicase-actions="selectedCases" multiple="true" refresh-callback="refresh"></ul>
+              </div>         
```
Extra space after after `div` should be removed.

deb1990 commented on ang/civicase/CaseListTable.js:
```diff
@@ -29,13 +30,98 @@
+     * Initiate subscribers
+     */
+    function initSubscribers () {
+      $scope.$on('bulkSelection', function (event, condition) {
```
Move the details to separate functions.
```
$scope.$on('bulkSelection', bulkSelectionEventListener);
$scope.$on('bulkSelectionCheckboxClicked', bulkSelectionCheckboxClickedListener);
```

deb1990 commented on ang/civicase/CaseListTable.js:
```diff
@@ -29,13 +30,98 @@
+          deselectAllCases();
+        } else if (condition === 'visible') {
+          selectDisplayedCases();
+        } else if (condition === 'all') {
```
`} else if (condition === 'all' && $scope.isSelectAllAvailable) {`

deb1990 commented on ang/civicase/CaseListTable.js:
```diff
@@ -29,13 +30,98 @@
+    }
+
+    /**
+     * Asynchronously get all cases for the select all bulk
```
`Asynchronously get all cases for the bulk actions select all functionality `

deb1990 commented on ang/civicase/CaseListTable.js:
```diff
@@ -29,13 +30,98 @@
+    }
+
+    /**
+     * Select all functionality
```
`Select All Cases`

deb1990 commented on ang/civicase/CaseListTable.js:
```diff
@@ -29,13 +30,98 @@
+    function selectDisplayedCases () {
+      _.each($scope.cases, function (item, index) {
+        $scope.cases[index].selected = true;
+        var isCurrentCaseInSelectedCases = _.find($scope.selectedCases, {
```
Variable should be declared outside of loop.

deb1990 commented on ang/civicase/CaseListTable.js:
```diff
@@ -29,13 +30,98 @@
+     */
+    function selectEveryCase () {
+      $scope.selectedCases = [];
+      $scope.selectedCases = _.each(allCases, formatCase);
```
I though we wont be fomatting all cases, as its a performance problem. Why is it necessary?

swastikpareek replied:
We need some keys like `item.status` that are created using formatCase Function. I don't think it is a big performance issue as this happens in parallel and is non blocking.

deb1990 commented on ang/civicase/CaseListTable.js:
```diff
@@ -29,13 +30,98 @@
+    /**
+     * Initiate subscribers
+     */
+    function initSubscribers () {
```
Private functions should be defined after public functions. JS refactoring wiki needs to be followed.

deb1990 commented on ang/civicase/CaseListTable.js:
```diff
@@ -210,13 +297,31 @@
+     */
+    function casesWatcher (cases) {
+      // if case is in selectedCases array update the UI model (checkbox)
+      _.each(cases, function (item, index) {
```
I think we discussed that we will be storing the IDs only. can we do that?

swastikpareek replied:
I think that we discussed it already that  we will be storing the full case object. As if we just store the keys, then we have to create new services for fetching the details of each case with an id and it will be a complex architecture and might introduce regressions.

swastikpareek commented on ang/civicase/BulkActionsCheckboxes.js:
```diff
@@ -0,0 +1,44 @@
+    $scope.showCheckboxes = false;
+
+    $scope.$watch('selectedItems', function () {
+      $rootScope.$broadcast('stickyCaseListHeaderReinitiatePos');
```
We can't use $emit as its subscribers will only be available inside its parent hierarchy, but where it is subscribers are added, are outside that scope.
So if we emit `stickyCaseListHeaderReinitiatePos` from `BulkActionsCheckboxes` it can't be caught in CaseListDirective as it doesn't come in `BulkActionsCheckboxes` parent Hierarchy.

Let me know if we need discussion on it.

deb1990 replied:
Yes, but if we always use `$rootScope.$watch` then this would work.

swastikpareek commented:
> 1.Actions text should not be ALL CAPS, also font color does not match design.

Fixed

> In Case Viewing mode, the Bulk Actions should be visible according to the design.

 We don't need to show the checkbox and the actions menu when the case is selected (Default behaviour of the old dcase site)

> As per discussion with Shitij, count of Everything, Displayed should be shown in brackets.

Added.

deb1990 commented on ang/civicase/BulkActionsCheckboxes.js:
```diff
@@ -0,0 +1,62 @@
+     * - 'visible' :  Selects all visible selections
+     * - 'none' : Deselects all
+     *
+     * @params {String} condition
```
Params are not updated.

deb1990 commented on ang/test/civicase/BulkActionsCheckboxes.spec.js:
```diff
@@ -0,0 +1,146 @@
+      element.isolateScope().$emit = originalEmitFunction;
+    });
+
+    describe('called for all selection', function () {
```
`when` prefix.

deb1990 commented on ang/test/civicase/BulkActionsCheckboxes.spec.js:
```diff
@@ -0,0 +1,146 @@
+      });
+
+      it('emits bulkSelection event with "all" parameter', function () {
+        expect(element.isolateScope().$emit).toHaveBeenCalledWith('bulkSelection', 'all');
```
The event name should follow the format, as per the previous comment.

deb1990 commented on ang/test/civicase/BulkActionsCheckboxes.spec.js:
```diff
@@ -0,0 +1,146 @@
+  });
+
+  /**
+   * Function responsible for setting up compilation of the directive
```
`Compile the directive`

deb1990 commented on scss/components/_bulk-actions.scss:
```diff
@@ -0,0 +1,71 @@
+  }
+}
+
+.civicase__bulkactions-checkbox-dropdown {
```
`civicase__bulkactions-select-mode-dropdown`

deb1990 commented on scss/components/_bulk-actions.scss:
```diff
@@ -0,0 +1,71 @@
+  vertical-align: middle;
+}
+
+.civicase__bulkactions-actions {
```
`civicase__bulkactions-actions-dropdown`

swastikpareek commented on ang/test/civicase/BulkActionsMessage.spec.js:
```diff
@@ -0,0 +1,143 @@
+/* eslint-env jasmine */
+
+describe('BulkActionsMessage', function () {
```
We are checking basic tests and the logical implementation inside the HTML template.

swastikpareek commented:
@deb1990 : We do need timeouts in the subscribers for `header-position` subscribers. The top offset calculations is done correctly when we use timeouts.
